odoo/http.py

```diff
@@ -1,836 +1,1081 @@
+    :type name: string
 
-    def redirect_query(self, location, query=None, code=303, local=True):
-        if query:
-            location += '?' + urls.url_encode(query)
-        return self.redirect(location, code=code, local=local)
+    :param value: the value to set for the `name` header
+    :type value: string
+
+    :return: the updated headers
+    :rtype: list of tuples (name, value)
+    """
+    dictheaders = dict(headers)
+    dictheaders[name] = value
+    return list(dictheaders.items())
+
+def set_safe_image_headers(headers, content):
+    """Return new headers based on `headers` but with `Content-Length` and
+    `Content-Type` set appropriately depending on the given `content` only if it
+    is safe to do, as well as `X-Content-Type-Options: nosniff` so that if the
+    file is of an unsafe type, it is not interpreted as that type if the
+    `Content-type` header was already set to a different mimetype"""
+    content_type = guess_mimetype(content)
+    safe_types = ['image/jpeg', 'image/png', 'image/gif', 'image/x-icon']
+    if content_type in safe_types:
+        headers = set_header_field(headers, 'Content-Type', content_type)
+    headers = set_header_field(headers, 'X-Content-Type-Options', 'nosniff')
+    headers = set_header_field(headers, 'Content-Length', len(content))
+    return headers
+
+
+#----------------------------------------------------------
+# Controller and routes
+#----------------------------------------------------------
+addons_manifest = {}  # TODO @juc, move to odoo.modules.module
+
+class Controller:
+    """
+    Class mixin that provide module controllers the ability to serve
+    content over http and to be extended in child modules.
+
+    Each class :ref:`inheriting <python:tut-inheritance>` from
+    :class:`~odoo.http.Controller` can use the :func:`~odoo.http.route`:
+    decorator to route matching incomming web requests to decorated
+    method.
+
+    Like models, controllers can be extended by other modules. The
+    extension mechanism is different as controllers can work in a
+    database-free environment thus lack the necessary to use the
+    :class:~odoo.api.Registry:.
+
+    To *override* a controller, :ref:`inherit <python:tut-inheritance>`
+    from its class, override relevant methods and re-expose them with
+    :func:`~odoo.http.route`:. Please note that the decorators of all
+    methods are combined, if the overriding method’s decorator has no
+    argument all previous ones will be kept, any provided argument will
+    override previously defined ones.
+
+        class GreetingController(odoo.http.Controller):
+            @route('/greet', type='http', auth='public')
+            def greeting(self):
+                return 'Hello'
+
+        class UserGreetingController(GreetingController):
+            @route(auth='user')  # override auth, keep path and type
+            def greeting(self):
+                return super().handler()
```

Comment:
Should be encapsulated in a code-block directive if you/we plan to include those docstrings in the doc later :).